File: src/main/java/nz/co/searchwellington/modification/ContentDeletionService.java
package nz.co.searchwellington.modification;

import nz.co.searchwellington.feeds.RssfeedNewsitemService;
import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.Newsitem;
import nz.co.searchwellington.model.PublishedResource;
import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.model.Watchlist;
import nz.co.searchwellington.model.Website;
import nz.co.searchwellington.repositories.HandTaggingDAO;
import nz.co.searchwellington.repositories.HibernateResourceDAO;
import nz.co.searchwellington.repositories.SupressionService;
import nz.co.searchwellington.repositories.TagDAO;
import nz.co.searchwellington.repositories.mongo.MongoSnapshotDAO;
import nz.co.searchwellington.repositories.solr.SolrQueryService;

import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class ContentDeletionService {
	
	private static Logger log = Logger.getLogger(ContentDeletionService.class);
    	
	private SupressionService supressionService;
	private RssfeedNewsitemService rssfeedNewsitemService;
	private HibernateResourceDAO resourceDAO;
	private MongoSnapshotDAO snapshotDAO;
	private SolrQueryService solrQueryService;
	private HandTaggingDAO handTaggingDAO;
	private TagDAO tagDAO;
	
	public ContentDeletionService() {
	}
	
	@Autowired
	public ContentDeletionService(SupressionService supressionService,
			RssfeedNewsitemService rssfeedNewsitemService,
			HibernateResourceDAO resourceDAO, MongoSnapshotDAO snapshotDAO,
			SolrQueryService solrQueryService, HandTaggingDAO handTaggingDAO,
			TagDAO tagDAO) {
		this.supressionService = supressionService;
		this.rssfeedNewsitemService = rssfeedNewsitemService;
		this.resourceDAO = resourceDAO;
		this.snapshotDAO = snapshotDAO;
		this.solrQueryService = solrQueryService;
		this.handTaggingDAO = handTaggingDAO;
		this.tagDAO = tagDAO;
	}

	@Transactional
	public void performDelete(Resource resource) {		
		handTaggingDAO.clearTags(resource);
		
		if (resource.getType().equals("W")) {
			removePublisherFromPublishersContent(resource);            	
		}
		
		if (resource.getType().equals("F")) {
			removeFeedFromFeedNewsitems((Feed) resource);			
			removeRelatedFeedFromTags((Feed) resource);
		}
		
		if (resource.getType().equals("N")) {
			log.info("Deleted item is a newsitem; checking if it's in an accepted feed.");
			Newsitem deletedNewsitem = (Newsitem) resource;
			if (rssfeedNewsitemService.isUrlInAcceptedFeeds(deletedNewsitem.getUrl())) {
				log.info("Supressing deleted newsitem url as it still visible in an automatically deleted feed: " + deletedNewsitem.getUrl());
				suppressDeletedNewsitem(deletedNewsitem);
			} else {
				log.info("Not found in live feeds; not supressing");
			}
		}
		
		snapshotDAO.evict(resource.getUrl());
		solrQueryService.deleteResourceFromIndex(resource.getId());
		resourceDAO.deleteResource(resource);
	}
	
	private void removeRelatedFeedFromTags(Feed editResource) {
		for (Tag tag : tagDAO.getAllTags()) {
			if (tag.getRelatedFeed() != null && tag.getRelatedFeed().equals(editResource)) {
				tag.setRelatedFeed(null);
			}
		}
	}
	
	private void suppressDeletedNewsitem(Newsitem deletedNewsitem) {
		log.info("Deleting a newsitem whose url still appears in a feed; suppressing the url: " + deletedNewsitem.getUrl());			
		supressionService.suppressUrl(deletedNewsitem.getUrl());
	}
	
	private void removePublisherFromPublishersContent(Resource editResource) {
    	Website publisher = (Website) editResource;
    	for (PublishedResource published : resourceDAO.getNewsitemsForPublishers(publisher)) {
    		published.setPublisher(null);
    		resourceDAO.saveResource(publisher);					
    	}
    	for (Feed feed : publisher.getFeeds()) {
    		feed.setPublisher(null);
    		resourceDAO.saveResource(feed);					
    	}
    	for (Watchlist watchlist : publisher.getWatchlist()) {
    		watchlist.setPublisher(null);
    		resourceDAO.saveResource(watchlist);					
    	}
    }
	
    private void removeFeedFromFeedNewsitems(Feed feed) {    	
    	for (Newsitem newsitem : resourceDAO.getNewsitemsForFeed(feed)) {
    		newsitem.setFeed(null);
    		resourceDAO.saveResource(newsitem);					
    	}    	
    }

}

File: src/main/scala/nz/co/searchwellington/modification/ContentDeletionService.scala
package nz.co.searchwellington.modification

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.model._
import nz.co.searchwellington.repositories.SuppressionDAO
import nz.co.searchwellington.repositories.elasticsearch.ElasticSearchIndexer
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.apache.commons.logging.LogFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Await, Future}

@Component class ContentDeletionService @Autowired()(suppressionDAO: SuppressionDAO,
                                                     mongoRepository: MongoRepository,
                                                     elasticSearchIndexer: ElasticSearchIndexer,
                                                     contentUpdateService: ContentUpdateService)
  extends ReasonableWaits {

  private val log = LogFactory.getLog(classOf[ContentDeletionService])

  def performDelete(resource: Resource): Boolean = {
    try {
      log.info("Deleting resource: " + resource)

      // Resource type specific actions
      resource match {
        case website: Website =>
          removePublisherFromPublishersContent(website)
        case feed: Feed =>
          Await.result(removeFeedFromFeedNewsitems(feed), OneMinute)
        case newsitem: Newsitem =>
          log.info("Suppressing deleted newsitem url to prevent it been reaccepted from a feed: " + newsitem.page)
          Await.result(suppressionDAO.addSuppression(newsitem.page), TenSeconds)
        case _ =>
      }

      Await.result(elasticSearchIndexer.deleteResource(resource._id).flatMap { dr =>
        mongoRepository.removeResource(resource)
      }, TenSeconds)
      true

    } catch {
      case e: Exception =>
        log.error("Delete error: " + e.getMessage, e)
        false
    }
  }

  private def removePublisherFromPublishersContent(publisher: Website): Unit = {
    mongoRepository.getResourcesIdsForPublisher(publisher).map { published =>
      published.foreach { publishedResourceId =>
        mongoRepository.getResourceByObjectId(publishedResourceId).map { maybeResource =>
          maybeResource.map {
            case published: PublishedResource =>
              log.info("Clearing publisher from: " + published.title)
              published.clearPublisher()
              contentUpdateService.update(published)
          }
        }
      }
    }
  }

  private def removeFeedFromFeedNewsitems(feed: Feed): Future[Seq[Boolean]] = {
    mongoRepository.getAllNewsitemsForFeed(feed).flatMap { ns =>
      Future.sequence {
        ns.map { n =>
          log.info("Removing feed from newsitem: " + n.title)
          val withoutFeed = n.copy(feed = None)
          contentUpdateService.update(withoutFeed)
        }
      }
    }
  }

}
